File: backend/service/payment/src/main/java/com/leopoldhsing/digitalhippo/payment/config/SnsConfig.java
package com.leopoldhsing.digitalhippo.payment.config;

import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.core.env.Environment;
import software.amazon.awssdk.auth.credentials.AwsBasicCredentials;
import software.amazon.awssdk.auth.credentials.StaticCredentialsProvider;
import software.amazon.awssdk.regions.Region;

import java.net.URI;

@Configuration
public class SnsConfig {

    private final String accessKey;
    private final String secretKey;
    private final String region;
    private final String snsEndpoint;

    public SnsConfig(Environment env) {
        this.accessKey = env.getProperty("spring.cloud.aws.credentials.access-key");
        this.secretKey = env.getProperty("spring.cloud.aws.credentials.secret-key");
        this.region = env.getProperty("spring.cloud.aws.region.static");
        this.snsEndpoint = env.getProperty("spring.cloud.aws.sns.endpoint");
    }

    @Bean
    public software.amazon.awssdk.services.sns.SnsClient getSnsClient() {
        return software.amazon.awssdk.services.sns.SnsClient.builder()
                .credentialsProvider(StaticCredentialsProvider.create(AwsBasicCredentials.create(accessKey, secretKey)))
                .endpointOverride(URI.create(snsEndpoint == null ? "http://localstack:4566" : snsEndpoint))
                .region(Region.of(region))
                .build();
    }
}

File: service/user/src/main/kotlin/com/leopoldhsing/digitalhippo/user/config/SnsConfig.kt
package com.leopoldhsing.digitalhippo.user.config

import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.core.env.Environment
import software.amazon.awssdk.auth.credentials.AwsBasicCredentials
import software.amazon.awssdk.auth.credentials.AwsCredentialsProvider
import software.amazon.awssdk.regions.Region
import software.amazon.awssdk.services.sns.SnsClient
import java.net.URI

@Configuration
open class SnsConfig(private val env: Environment) {

    private val accessKey: String? = env.getProperty("spring.cloud.aws.credentials.access-key")
    private val secretKey: String? = env.getProperty("spring.cloud.aws.credentials.secret-key")
    private val region: String? = env.getProperty("spring.cloud.aws.region.static")
    private var snsEndpoint: String? = env.getProperty("spring.cloud.aws.sns.endpoint")

    @Bean
    open fun getSnaClient(): SnsClient {
        return SnsClient.builder()
            .credentialsProvider(AwsCredentialsProvider { AwsBasicCredentials.create(accessKey, secretKey) })
            .endpointOverride(URI.create(if (snsEndpoint == null) "http://localstack:4566" else snsEndpoint!!))
            .region(Region.of(region))
            .build()
    }
}